File: packages/typescript/src/signals.ts
/** Structured signal envelope types (#1098). */

export const StructuredSignalEventType = {
  TRADE_IDEA: 'trade_idea',
  BUY: 'buy',
  SCALE_IN: 'scale_in',
  SELL: 'sell',
  PARTIAL_SELL: 'partial_sell',
  AMEND: 'amend',
  BREAKEVEN: 'breakeven',
  CANCEL: 'cancel',
  POSITION_UPDATE: 'position_update',
  TRADE_REVIEW: 'trade_review',
} as const;

export type StructuredSignalEventType =
  (typeof StructuredSignalEventType)[keyof typeof StructuredSignalEventType];

export const StructuredSignalChain = {
  SOLANA: 'solana',
  HYPERLIQUID: 'hyperliquid',
} as const;

export type StructuredSignalChain =
  (typeof StructuredSignalChain)[keyof typeof StructuredSignalChain];

export const StructuredSignalSide = {
  LONG: 'long',
  SHORT: 'short',
} as const;

export type StructuredSignalSide =
  (typeof StructuredSignalSide)[keyof typeof StructuredSignalSide];

export const StructuredSignalTradeType = {
  SPOT: 'spot',
  PERP: 'perp',
  VIRTUAL: 'virtual',
} as const;

export type StructuredSignalTradeType =
  (typeof StructuredSignalTradeType)[keyof typeof StructuredSignalTradeType];

export type InboundSignalBody = Record<string, unknown> & {
  text?: string;
  idempotencyKey?: string;
  eventType?: StructuredSignalEventType;
  chain?: StructuredSignalChain;
  symbol?: string;
  side?: StructuredSignalSide;
  tradeType?: StructuredSignalTradeType;
  amount?: number;
  reasoning?: string;
  confidence?: number;
  positionRef?: string;
  action?: 'buy' | 'sell';
  tokenAddress?: string;
};

export type AgentSignalResponse = {
  outcome?: 'matched' | 'unmatched' | 'skipped' | 'error';
  signalId?: string;
  status?: string;
  skipReason?: string;
};

export type OutboundExecutionWebhook = {
  event: 'signal.execution' | 'signal.execution.failed';
  signalId: string;
  developerAgentId: string;
  status: string;
  executionResult?: {
    success: boolean;
    txHash?: string;
    executedAmountUsd?: number;
    errorMessage?: string;
    executedAt?: string;
  };
  timestamp: string;
};

File: packages/python/src/echozero/signals.py
"""Structured signal envelope types (#1098)."""

from __future__ import annotations

from typing import Literal, TypedDict

StructuredSignalEventType = Literal[
    "trade_idea",
    "buy",
    "scale_in",
    "sell",
    "partial_sell",
    "amend",
    "breakeven",
    "cancel",
    "position_update",
    "trade_review",
]

StructuredSignalChain = Literal["solana", "hyperliquid"]
StructuredSignalSide = Literal["long", "short"]
StructuredSignalTradeType = Literal["spot", "perp", "virtual"]


class AgentSignalResponse(TypedDict, total=False):
    outcome: Literal["matched", "unmatched", "skipped", "error"]
    signalId: str
    status: str
    skipReason: str


class OutboundExecutionWebhook(TypedDict, total=False):
    event: Literal["signal.execution", "signal.execution.failed"]
    signalId: str
    developerAgentId: str
    status: str
    executionResult: dict[str, object]
    timestamp: str
